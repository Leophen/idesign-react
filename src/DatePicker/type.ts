export interface DatePickerProps {
  /**
   * 类名
   */
  className?: string;
  /**
   * 自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 选中值
   */
  value?: string;
  /**
   * 默认选中值，非受控属性
   */
  defaultValue?: string;
  /**
   * 触发方式
   * @default click
   */
  trigger?: 'hover' | 'click' | 'context-menu';
  /**
   * 是否禁用选择器
   * @default false
   */
  disabled?: boolean;
  /**
   * 日期选择器类型
   * @default date
   */
  type?: 'date' | 'range';
  /**
   * 占位符
   * @default date
   */
  placeholder?: 'date' | 'range';
  /**
   * 范围日期分隔符
   * @default -
   */
  rangeSeparator?: string;
  /**
   * 每周的第一天，0 为周日，1 为周一
   * @default 1
   */
  firstDayOfWeek?: 0 | 1;
  /**
   * 选中值发生变化时触发
   */
  onChange?: (value: string) => void;
}

export interface DateHeaderBtnProps {
  icon?: string;
  disabled?: boolean;
  onClick?: Function;
}

export interface DatePanelProps {
  value?: string;
  defaultDisplayVal?: string;
  rangeValue?: string[];
  firstDayOfWeek?: number;
  onChange?: Function;
  onHover?: Function;
  onClose?: Function
}

export interface DatePanelRangeProps extends DatePanelProps { }

export interface MonthSelectPanelProps {
  month: number;
  onChange: Function;
}

export interface YearSelectPanelProps {
  year: number;
  onChange: Function;
}
