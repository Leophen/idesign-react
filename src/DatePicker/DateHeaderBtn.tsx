import React from 'react';
import classNames from 'classnames';
import './index.scss';
import Icon from '../Icon';
import { DateHeaderBtnProps } from './type';

const DateHeaderBtn: React.FC<DateHeaderBtnProps> = (props) => {
  const {
    icon = "ArrowDoubleLeft",
    disabled = false,
    onClick = () => { }
  } = props

  const handleClickBtn = () => {
    !disabled && onClick?.()
  }

  return (
    <div
      className={classNames(
        'i-date-panel__header-icon',
        disabled && 'i-date-panel__header-icon__disabled'
      )}
      onClick={handleClickBtn}
    >
      <Icon name={icon} />
    </div>
  )
}

export default DateHeaderBtn;
