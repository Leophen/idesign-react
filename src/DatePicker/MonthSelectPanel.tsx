import React, { useState, useEffect } from 'react';
import classNames from 'classnames';
import './index.scss';
import _ from 'lodash';
import Icon from '../Icon';
import { MonthSelectPanelProps } from './type';

const MonthSelectPanel: React.FC<MonthSelectPanelProps> = (props) => {
  const {
    month = 6,
    onChange
  } = props

  const [innerMonth, setInnerMonth] = useState(month)

  useEffect(() => {
    setInnerMonth(month)
  }, [month])

  const handleLastMonth = (e: React.MouseEvent) => {
    e.stopPropagation()
    innerMonth > 0 && setInnerMonth(innerMonth - 1)
  }

  const handleNextMonth = (e: React.MouseEvent) => {
    e.stopPropagation()
    innerMonth < 11 && setInnerMonth(innerMonth + 1)
  }

  const handleClickItem = (e: React.MouseEvent, item: number) => {
    e.stopPropagation()
    onChange?.(item)
  }

  return (
    <div className="i-date__month-select-panel" onClick={(e) => e.stopPropagation()}>
      <header className="i-date__month-select-title">
        <div className="i-date-panel__header-icon" onClick={handleLastMonth}>
          <Icon name="ArrowLeft" />
        </div>
        <div className="i-date-panel__header-title">
          {innerMonth + 1} æœˆ
        </div>
        <div className="i-date-panel__header-icon" onClick={handleNextMonth}>
          <Icon name="ArrowRight" />
        </div>
      </header>
      <section className="i-date__month-select-list">
        {_.range(0, 12).map(item => (
          <div
            className={classNames(
              'i-date__month-select-item',
              innerMonth === item && 'i-date__month-select-item__selected',
              new Date().getMonth() === item && 'i-date__month-select-item__current'
            )}
            key={item}
            onClick={(e) => handleClickItem(e, item)}
          >
            {item + 1}
          </div>
        ))}
      </section>
    </div>
  )
}

export default MonthSelectPanel;
