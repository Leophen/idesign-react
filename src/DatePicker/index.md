---
nav:
  title: 组件
  path: /components
group:
  title: 表单组件
  order: 3
order: 8
---

# DatePicker 日期选择器

用于选择或输入日期。

## 基本用法

```tsx
import React, { useState } from 'react'
import { DatePicker } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState('2022-05-20')

  const handleChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <>
      <h4>无默认值</h4>
      <DatePicker />
      <h4>有默认值</h4>
      <DatePicker defaultValue={value} />
      <h4>有受控值</h4>
      <DatePicker value={value} />
      <h4>一般用法</h4>
      <DatePicker
        value={value}
        onChange={handleChange}
      />
    </>
  )
}

export default App
```

## 日期范围选择器

可通过 `type` 属性指定为日期范围选择器。

```tsx
import React, { useState } from 'react'
import { DatePicker } from 'idesign-react'

const App = () => {
  const [value, setValue] = useState(['2022-05-20', '2022-11-11'])

  const handleChange = (val) => {
    console.log(val)
    setValue(val)
  }

  return (
    <>
      <h4>无默认值</h4>
      <DatePicker type="range" />
      <h4>有默认值</h4>
      <DatePicker type="range" defaultValue={value} />
      <h4>有受控值</h4>
      <DatePicker type="range" value={value} />
      <h4>一般用法</h4>
      <DatePicker
        type="range"
        value={value}
        onChange={handleChange}
      />
    </>
  )
}

export default App
```

## 禁用状态

可通过 `disabled` 属性控制为禁用状态。

```tsx
import React from 'react'
import { DatePicker } from 'idesign-react'

const App = () => {
  return (
    <DatePicker disabled />
  )
}

export default App
```

## 不同触发方式

可通过 `trigger` 属性指定下拉菜单触发方式，默认为 `click`。

```tsx
import React from 'react'
import { DatePicker } from 'idesign-react'

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <DatePicker trigger="click" />
      <DatePicker trigger="hover" />
      <DatePicker trigger="context-menu" />
    </div>
  )
}

export default App
```

## 每周的第一天

可通过 `firstDayOfWeek` 属性设置每周的第一天，`0` 为周日，`1` 为周一。

```tsx
import React from 'react'
import { DatePicker } from 'idesign-react'

const App = () => {
  return (
    <div className="idesign-demo-block-row">
      <DatePicker firstDayOfWeek={0} />
      <DatePicker firstDayOfWeek={1} />
    </div>
  )
}

export default App
```

## 自定义分隔符

可通过 `rangeSeparator` 属性指定日期范围选择器的连接分隔符。

```tsx
import React from 'react'
import { DatePicker } from 'idesign-react'

const App = () => {
  return (
    <DatePicker type="range" rangeSeparator="/" />
  )
}

export default App
```

<API />
