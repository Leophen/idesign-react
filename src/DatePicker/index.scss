.i-date-picker {
  &[data-type='range'] {
    .i-input {
      min-width: 196px;
    }
  }
}

.i-date-panel {
  box-sizing: border-box;
  width: auto;
}

.i-date-panel__header {
  position: relative;
  display: flex;
  align-items: center;
  justify-content: space-between;
  box-sizing: border-box;
  width: 100%;
  height: 50px;
  padding: 12px 16px 12px 20px;
}

.i-date-panel__header-today {
  box-sizing: border-box;
  width: 12px;
  height: 12px;
  margin: 0 6px;
  border: 1px solid var(--i-font-2);
  border-radius: 50%;
  cursor: pointer;
  transition: all 0.1s ease-out;
  &:hover {
    border-color: var(--i-primary-txt);
  }
}

.i-date-panel__header-handle {
  display: flex;
  gap: 2px;
  align-items: center;
}

.i-date-panel__header-icon {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 50%;
  cursor: pointer;
  transition: all 0.1s ease-out;
  .i-icon {
    margin: 0;
  }
  &:hover {
    background: var(--i-bg-hover);
  }
  &.i-date-panel__header-icon__disabled {
    cursor: not-allowed;
    .i-icon {
      color: var(--i-font-disabled);
    }
    &:hover {
      background: unset;
    }
  }
}

.i-date-panel__header-title-wrapper {
  display: flex;
  gap: 2px;
  align-items: center;
}

.i-date-panel__header-title {
  cursor: pointer;
  transition: all 0.1s ease-out;
  user-select: none;
  &:hover {
    color: var(--i-primary-txt);
  }
}

.i-date__year-select-panel,
.i-date__month-select-panel {
  box-sizing: border-box;
  width: auto;
  padding: 8px 12px 16px;
}

.i-date__year-select-title,
.i-date__month-select-title {
  display: flex;
  align-items: center;
  justify-content: space-between;
  box-sizing: border-box;
  width: 100%;
  padding-bottom: 8px;
  border-bottom: 1px solid var(--i-border);
}

.i-date__year-select-list,
.i-date__month-select-list {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  gap: 8px;
  padding-top: 16px;
}

.i-date__year-select-item,
.i-date__month-select-item {
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  width: 42px;
  height: 26px;
  font-size: 12px;
  border-radius: 16px;
  cursor: pointer;
  transition: all 0.1s ease-out;
  user-select: none;
  &.i-date__year-select-item__current,
  &.i-date__month-select-item__current {
    color: var(--i-primary-txt);
    background: var(--i-primary-bg);
    &:hover {
      color: var(--i-primary-txt);
    }
  }
  &.i-date__year-select-item__selected,
  &.i-date__month-select-item__selected {
    color: #fff;
    background: var(--i-primary);
    &:hover {
      color: #fff;
    }
  }
  &:hover {
    color: var(--i-primary-txt);
  }
}

.i-date-panel__content {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  padding: 0 12px 12px;
}

.i-date-panel__week-wrapper,
.i-date-panel__day-wrapper {
  display: flex;
  flex-wrap: wrap;
  width: 252px;
  row-gap: 4px;
}

.i-date-panel__week-item,
.i-date-panel__day-item {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 36px;
  height: 36px;
}

.i-date-panel__day-item {
  position: relative;
  cursor: pointer;
  user-select: none;
  &.i-date-panel__day-item__current {
    color: var(--i-primary-txt);
    .i-date-panel__day-item-txt {
      background: var(--i-primary-bg);
    }
    &:hover {
      color: var(--i-primary-txt);
    }
  }
  &.i-date-panel__day-item__in-range {
    color: #fff;
    background: var(--i-primary-disabled);
    .i-date-panel__day-item-txt {
      background: unset;
    }
    &.-range__start {
      border-radius: 50% 0 0 50%;
    }
    &.-range__end {
      border-radius: 0 50% 50% 0;
    }
    &.-range__same {
      border-radius: 50%;
    }
  }
  &.i-date-panel__day-item__selected {
    color: #fff;
    .i-date-panel__day-item-txt {
      background: var(--i-primary);
    }
    &:hover {
      color: #fff;
    }
  }
  &.i-date-panel__day-item__blur {
    color: var(--i-font-5);
    .i-date-panel__day-item-txt {
      background: unset;
    }
    &:hover {
      color: var(--i-font-5);
    }
  }
  &:hover {
    color: var(--i-primary-txt);
  }
}

.i-date-panel__day-item-txt {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 30px;
  height: 30px;
  color: inherit;
  border-radius: 50%;
}

.i-date-panel__range {
  display: flex;
  align-items: center;
  .i-date-panel:last-of-type {
    border-left: 1px solid var(--i-border);
  }
}
