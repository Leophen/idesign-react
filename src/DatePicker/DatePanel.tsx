import React, { useState, useEffect } from 'react';
import classNames from 'classnames';
import './index.scss';
import _ from 'lodash';
import dayjs from 'dayjs';
import isBetween from 'dayjs/plugin/isBetween'
import Popup from '../Popup';
import YearSelectPanel from './YearSelectPanel';
import MonthSelectPanel from './MonthSelectPanel';
import DateHeaderBtn from './DateHeaderBtn';
import { DATE_FORMAT, WEEK_HEAD_0, WEEK_HEAD_1 } from './index'
import { DatePanelProps } from './type';

dayjs.extend(isBetween)

const DatePanel: React.FC<DatePanelProps> = (props) => {
  const {
    firstDayOfWeek,
    value,
    rangeValue,
    defaultDisplayVal,
    onChange,
    onHover,
    onClose
  } = props;

  const [dateVal, setDateVal] = useState(value)
  const [displayVal, setDisplayVal] = useState(value)
  useEffect(() => {
    if (defaultDisplayVal) {
      setDisplayVal(defaultDisplayVal)
    } else {
      setDisplayVal(value)
    }
  }, [value])

  const handleSwitchDate = (mode: 'add' | 'sub', type: 'year' | 'month', step = 1) => {
    let newDateVal
    if (mode === 'add') {
      newDateVal = dayjs(displayVal).add(step, type).format(DATE_FORMAT)
    } else {
      newDateVal = dayjs(displayVal).subtract(step, type).format(DATE_FORMAT)
    }
    setDisplayVal(newDateVal)
  }

  const handleToday = () => {
    setDisplayVal(dayjs().format(DATE_FORMAT))
  }

  const [yearPanelVisible, setYearPanelVisible] = useState(false)
  const [monthPanelVisible, setMonthPanelVisible] = useState(false)

  const handleSelectYear = (val: number) => {
    const newDateVal = dayjs(displayVal).year(val).format(DATE_FORMAT)
    setDisplayVal(newDateVal)
    setYearPanelVisible(false)
  }

  const handleSelectMonth = (val: number) => {
    const newDateVal = dayjs(displayVal).month(val).format(DATE_FORMAT)
    setDisplayVal(newDateVal)
    setMonthPanelVisible(false)
  }

  const handleSelectDay = (type: string, val: string) => {
    value && type !== 'current' && setDisplayVal(val)
    value && setDateVal(val)
    onChange?.(val)
    onClose?.()
  }

  const handleEnterMonth = (val: string) => {
    onHover?.(val)
  }

  const getMonthPanelDays = () => {
    const currentMonthStart = dayjs(displayVal).startOf('month').day()
    let startIndex = 0
    // 每周第一天
    if (firstDayOfWeek === 1) {
      currentMonthStart === 0 ? (startIndex = 6) : (startIndex = currentMonthStart - 1)
    } else {
      startIndex = currentMonthStart
    }

    const panelDays = []

    const lastMonthDays = dayjs(displayVal).subtract(1, 'month').daysInMonth()
    for (let i = lastMonthDays - startIndex + 1; i < lastMonthDays + 1; i++) {
      panelDays.push({
        label: i,
        type: 'last',
        value: dayjs(displayVal).subtract(1, 'month').set('date', i).format(DATE_FORMAT)
      })
    }

    const currentMonthDays = dayjs(displayVal).daysInMonth()
    const currentDayVal = (index: number) => dayjs(displayVal).set('date', index).format(DATE_FORMAT)
    const rangeValid = rangeValue?.length === 2
    for (let i = 1; i < currentMonthDays + 1; i++) {
      panelDays.push({
        label: i,
        type: 'current',
        range: rangeValid && (dayjs(currentDayVal(i)).isBetween(dayjs(rangeValue[0]), dayjs(rangeValue[1]), "day") || currentDayVal(i) === rangeValue[0] || currentDayVal(i) === rangeValue[1]),
        rangeStart: rangeValid && currentDayVal(i) === rangeValue[0],
        rangeSame: rangeValid && rangeValue[0] === rangeValue[1],
        rangeEnd: rangeValid && currentDayVal(i) === rangeValue[1],
        value: currentDayVal(i)
      })
    }

    const remainLen = 42 - startIndex + 1 - currentMonthDays
    for (let i = 1; i < remainLen; i++) {
      panelDays.push({
        label: i,
        type: 'next',
        value: dayjs(displayVal).add(1, 'month').set('date', i).format(DATE_FORMAT)
      })
    }

    return panelDays
  }

  return (
    <div
      className="i-date-panel"
    >
      <header className="i-date-panel__header">
        <section className="i-date-panel__header-title-wrapper">
          <Popup
            visible={yearPanelVisible}
            content={
              <YearSelectPanel year={dayjs(displayVal).year()} onChange={handleSelectYear} />
            }
            trigger="click"
            onTrigger={visible => setYearPanelVisible(visible)}
          >
            <div className="i-date-panel__header-title">
              {`${dayjs(displayVal).year()}年`}
            </div>
          </Popup>
          <Popup
            visible={monthPanelVisible}
            content={
              <MonthSelectPanel month={dayjs(displayVal).month()} onChange={handleSelectMonth} />
            }
            trigger="click"
            onTrigger={visible => setMonthPanelVisible(visible)}
          >
            <div className="i-date-panel__header-title">
              {`${dayjs(displayVal).month() + 1}月`}
            </div>
          </Popup>
        </section>

        <section className="i-date-panel__header-handle">
          <DateHeaderBtn
            icon="ArrowDoubleLeft"
            disabled={false}
            onClick={() => handleSwitchDate('sub', "year")}
          />
          <DateHeaderBtn
            icon="ArrowLeft"
            disabled={false}
            onClick={() => handleSwitchDate('sub', "month")}
          />
          <div
            className="i-date-panel__header-today"
            onClick={handleToday}
          />
          <DateHeaderBtn
            icon="ArrowRight"
            disabled={false}
            onClick={() => handleSwitchDate('add', "month")}
          />
          <DateHeaderBtn
            icon="ArrowDoubleRight"
            disabled={false}
            onClick={() => handleSwitchDate('add', "year")}
          />
        </section>
      </header>

      <div className="i-date-panel__content">
        <header className="i-date-panel__week-wrapper">
          {firstDayOfWeek === 0 ?
            WEEK_HEAD_0.map(item => (
              <div className="i-date-panel__week-item" key={item}>
                {item}
              </div>
            )) :
            WEEK_HEAD_1.map(item => (
              <div className="i-date-panel__week-item" key={item}>
                {item}
              </div>
            ))}
        </header>
        <section className="i-date-panel__day-wrapper">
          {getMonthPanelDays().map(item => (
            <div
              className={classNames(
                'i-date-panel__day-item',
                item.range && 'i-date-panel__day-item__in-range',
                item.rangeStart && '-range__start',
                item.rangeSame && '-range__same',
                item.rangeEnd && '-range__end',
                item.type !== 'current' && 'i-date-panel__day-item__blur',
                (item.value === dateVal || rangeValue?.includes(item.value)) && 'i-date-panel__day-item__selected',
                item.value === dayjs().format(DATE_FORMAT) && 'i-date-panel__day-item__current'
              )}
              key={item.value}
              onClick={() => handleSelectDay(item.type, item.value)}
              onMouseEnter={() => handleEnterMonth(item.value)}
            >
              <div
                className='i-date-panel__day-item-txt'
              >
                {item.label}
              </div>
            </div>
          ))}
        </section>
      </div>
    </div>
  )
}

export default DatePanel;
