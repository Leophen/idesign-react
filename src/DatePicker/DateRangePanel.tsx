import React, { useState, useRef } from 'react';
import './index.scss';
import _ from 'lodash';
import dayjs from 'dayjs';
import isBetween from 'dayjs/plugin/isBetween'
import DatePanel from './DatePanel';
import { DATE_FORMAT } from './index'
import { DatePanelRangeProps } from './type';

dayjs.extend(isBetween)

const DateRangePanel: React.FC<DatePanelRangeProps> = (props) => {
  const {
    rangeValue = [dayjs().format(DATE_FORMAT), dayjs().add(1, 'month').format(DATE_FORMAT)],
    firstDayOfWeek,
    onChange,
    onClose
  } = props;

  const [innerRangeValue, setInnerRangeValue] = useState(rangeValue)
  const [rangeLock, setRangeLock] = useState(true)

  const handleSelect = (val: string) => {
    if (rangeLock) {
      // 滑选未解锁 -> 解锁
      if (innerRangeValue.length === 2) {
        hoverStartVal.current = val
        const newRangeValue = [val, val]
        setInnerRangeValue([...newRangeValue])
        setRangeLock(false)
      }
    } else {
      // 滑选已解锁 -> 锁定
      onChange?.(innerRangeValue)
      onClose?.()
      setRangeLock(true)
    }
  }

  const valToNum = (val: string) => Number(val.replaceAll('-', ''))

  const hoverStartVal = useRef('')
  const handleHover = (val: string) => {
    if (!rangeLock) {
      let newRangeValue = innerRangeValue
      if (valToNum(val) > valToNum(hoverStartVal.current)) {
        newRangeValue[0] = hoverStartVal.current
        newRangeValue[1] = val
      } else if (valToNum(val) === valToNum(hoverStartVal.current)) {
        newRangeValue[0] = val
        newRangeValue[1] = val
      } else {
        newRangeValue[0] = val
        newRangeValue[1] = hoverStartVal.current
      }
      setInnerRangeValue([...newRangeValue])
    }
  }

  return (
    <div
      className="i-date-panel__range"
    >
      <DatePanel
        firstDayOfWeek={firstDayOfWeek}
        rangeValue={innerRangeValue}
        defaultDisplayVal={innerRangeValue[0]}
        onChange={handleSelect}
        onHover={handleHover}
      />
      <DatePanel
        firstDayOfWeek={firstDayOfWeek}
        rangeValue={innerRangeValue}
        defaultDisplayVal={innerRangeValue[1]}
        onChange={handleSelect}
        onHover={handleHover}
      />
    </div>
  )
}

export default DateRangePanel;
