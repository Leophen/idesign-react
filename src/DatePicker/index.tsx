import React, { useState } from 'react';
import classNames from 'classnames';
import './index.scss';
import _ from 'lodash';
import dayjs from 'dayjs';
import isBetween from 'dayjs/plugin/isBetween'
import Popup from '../Popup';
import Input from '../Input';
import useDefault from '../hooks/useDefault';
import DatePanel from './DatePanel';
import DateRangePanel from './DateRangePanel';
import { DatePickerProps } from './type';

dayjs.extend(isBetween)

export const DATE_FORMAT = 'YYYY-MM-DD';
export const WEEK_HEAD_0 = ['日', '一', '二', '三', '四', '五', '六']
export const WEEK_HEAD_1 = ['一', '二', '三', '四', '五', '六', '日']

const getDefaultVal = (type: string) => {
  if (type === 'range') {
    return [dayjs().format(DATE_FORMAT), dayjs().add(1, 'month').format(DATE_FORMAT)]
  } else {
    return dayjs().format(DATE_FORMAT)
  }
}

const DatePicker: React.FC<DatePickerProps> = (props) => {
  const {
    className,
    style,
    value,
    defaultValue = getDefaultVal(props.type || 'date'),
    trigger = "click",
    disabled = false,
    type = 'date',
    rangeSeparator = '~',
    firstDayOfWeek = 1,
    onChange
  } = props;

  const [popupVisible, setPopupVisible] = useState(false)
  const switchPopup = (visible: boolean) => {
    setPopupVisible(visible)
  }

  const [innerValue, setInnerValue] = useDefault(value, defaultValue, onChange);

  const handleChange = (val: string) => {
    setInnerValue(val)
  }

  const handleClose = () => {
    setPopupVisible(false)
  }

  return (
    <div
      className={classNames(
        'i-date-picker',
        className
      )}
      data-type={type}
      style={{ ...style }}
    >
      <Popup
        content={
          type === 'date' ? (
            <DatePanel
              value={innerValue as string}
              firstDayOfWeek={firstDayOfWeek}
              onChange={handleChange}
              onClose={handleClose}
            />
          ) : (
            <DateRangePanel
              rangeValue={innerValue as string[]}
              firstDayOfWeek={firstDayOfWeek}
              onChange={handleChange}
              onClose={handleClose}
            />
          )
        }
        placement="bottom"
        trigger={trigger}
        className="i-date-popup"
        visible={popupVisible}
        disabled={disabled}
        onTrigger={switchPopup}
      >
        <Input
          readonly
          disabled={disabled}
          value={!_.isArray(innerValue) ? innerValue : `${innerValue[0]} ${rangeSeparator} ${innerValue[1]}`}
          placeholder="请选择日期"
          suffixIcon="Calendar"
        />
      </Popup>
    </div>
  );
};

DatePicker.displayName = 'DatePicker';

export default DatePicker;
