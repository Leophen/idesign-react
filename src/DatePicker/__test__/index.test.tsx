import { render } from '@testing-library/react';
import DatePicker from '../index';

describe('DatePicker 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(<DatePicker />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('value', () => {
    const { container } = render(<DatePicker value='2022-05-20' />);
    expect(container.firstChild.firstChild.firstChild.firstChild.value).toEqual('2022-05-20');
  });

  it('defaultValue', () => {
    const { container } = render(<DatePicker defaultValue='2022-05-20' />);
    expect(container.firstChild.firstChild.firstChild.firstChild.value).toEqual('2022-05-20');
  });

  it('disabled', () => {
    const { container } = render(<DatePicker disabled />);
    expect(container.firstChild.firstChild.firstChild).toHaveClass('i-input-is-disabled');
  });

  it('type', () => {
    const { container } = render(<DatePicker type="range" value={['2022-05-20', '2022-11-11']} />);
    expect(container.firstChild.firstChild.firstChild.firstChild.value).toEqual('2022-05-20 ~ 2022-11-11');
  });

  it('rangeSeparator', () => {
    const { container } = render(<DatePicker type="range" value={['2022-05-20', '2022-11-11']} rangeSeparator="/" />);
    expect(container.firstChild.firstChild.firstChild.firstChild.value).toEqual('2022-05-20 / 2022-11-11');
  });
});

