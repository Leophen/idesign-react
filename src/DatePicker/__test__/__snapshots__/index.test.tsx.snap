// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-date-picker"
    data-type="date"
  >
    <div
      class="i-popup__reference i-date-popup"
    >
      <div
        class="i-input i-input-is-readonly"
      >
        <input
          class="i-input__inner"
          placeholder="请选择日期"
          readonly=""
          value="2022-09-19"
        />
        <i
          class="i-icon icon-Calendar i-input-suffix-icon"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
